pkg/discovery-manager/discovery_controller.go

```diff
@@ -47,13 +56,309 @@ type DiscoveryReconciler struct {
+				discovery.Status.Attempt = 0
+				setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseFailed)
+			}
+		} else {
+			// Update the Discovery with the PeeringCandidate
+			discovery.Spec.PeeringCandidate = nodecorev1alpha1.GenericRef{
+				Name:      peeringCandidate.Name,
+				Namespace: DEFAULT_NAMESPACE,
+			}
+			if err := r.Update(ctx, &discovery); err != nil {
+				klog.Errorf("Error when updating Discovery %s: %s", discovery.Name, err)
+				return ctrl.Result{}, err
+			}
+			klog.Infof("Discovery %s updated", discovery.Name)
+			setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseSolved)
+		}
+		if err := r.updateSolverStatus(ctx, &discovery); err != nil {
+			klog.Errorf("Error when updating Solver %s status: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		} else {
+			/// TODO: for Ale, move this function to the solver controller
+			reservationCR := createReservationCustomResource(*peeringCandidate)
+			err := r.Create(context.Background(), reservationCR)
+			if err != nil {
+				klog.Errorf("Error when creating Reservation: %s", err)
+				setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseFailed)
+			} else {
+				klog.Infof("Reservation %s created", reservationCR.Name)
+			}
+			return ctrl.Result{}, err
+		}
+	case nodecorev1alpha1.PhaseSolved:
+
+		klog.Infof("Discovery %s solved", discovery.Name)
+		setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseSolved)
+		if err := r.updateSolverStatus(ctx, &discovery); err != nil {
+			klog.Errorf("Error when updating Solver %s status: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	case nodecorev1alpha1.PhaseFailed:
+		klog.Infof("Discovery %s failed, no Peering Candidate found", discovery.Name)
+		setDiscoveryPhase(&discovery, nodecorev1alpha1.PhaseFailed)
+		if err := r.updateSolverStatus(ctx, &discovery); err != nil {
+			klog.Errorf("Error when updating Solver %s status: %s", req.NamespacedName, err)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
 
 	return ctrl.Result{}, nil
 }
 
+// setDiscoveryPhase sets the phase of the discovery
+func setDiscoveryPhase(d *advertisementv1alpha1.Discovery, phase nodecorev1alpha1.Phase) {
+	d.Status.Phase.Phase = phase
+	d.Status.Phase.LastChangeTime = time.Now().Format(time.RFC3339)
+}
+
+// updateSolverStatus updates the status of the discovery
+func (r *DiscoveryReconciler) updateSolverStatus(ctx context.Context, discovery *advertisementv1alpha1.Discovery) error {
+	return r.Status().Update(ctx, discovery)
+}
+
+// getFlavour is a function that returns an array of Flavour that fit the Selector by performing a get request to an http server
+func getFlavour(selector nodecorev1alpha1.FlavourSelector) ([]*nodecorev1alpha1.Flavour, error) {
+	// Marshal the selector into JSON bytes
+	cpuInt, _ := selector.Cpu.AsInt64()
+	ramInt, _ := selector.Memory.AsInt64()
+	ephStorage, _ := selector.EphemeralStorage.AsInt64()
```

Comment:
this can become a standard parser that you can put inside the generic utils and use everywhere you need, no?

Reply:
Yessir, fixed!